File: qqbot-runtime/src/main/java/com/mieai/qqbot/runtime/security/KeyMaterial.java
package com.mieai.qqbot.runtime.security;

import java.util.Arrays;
import java.util.Objects;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;

/** Identified AES-256 master key supplied by a {@link KeyProvider}. */
public final class KeyMaterial {
    private static final int AES_256_BYTES = 32;

    private final String keyId;
    private final SecretKey key;

    public KeyMaterial(String keyId, SecretKey key) {
        this.keyId = requireKeyId(keyId);
        Objects.requireNonNull(key, "key must not be null");
        if (!"AES".equalsIgnoreCase(key.getAlgorithm())) {
            throw new IllegalArgumentException("key algorithm must be AES");
        }

        byte[] encoded = key.getEncoded();
        if (encoded == null) {
            throw new IllegalArgumentException("key must expose RAW key material");
        }
        try {
            if (encoded.length != AES_256_BYTES) {
                throw new IllegalArgumentException("key must contain exactly 256 bits");
            }
            this.key = new SecretKeySpec(encoded, "AES");
        } finally {
            Arrays.fill(encoded, (byte) 0);
        }
    }

    public String keyId() {
        return keyId;
    }

    SecretKey key() {
        return key;
    }

    @Override
    public String toString() {
        return "KeyMaterial[keyId=" + keyId + ", key=<redacted>]";
    }

    private static String requireKeyId(String value) {
        Objects.requireNonNull(value, "keyId must not be null");
        if (value.isBlank()) {
            throw new IllegalArgumentException("keyId must not be blank");
        }
        if (!value.equals(value.strip())) {
            throw new IllegalArgumentException("keyId must not have surrounding whitespace");
        }
        if (value.codePoints().anyMatch(Character::isWhitespace)) {
            throw new IllegalArgumentException("keyId must not contain whitespace");
        }
        if (value.codePoints().anyMatch(Character::isISOControl)) {
            throw new IllegalArgumentException("keyId must not contain control characters");
        }
        return value;
    }
}

File: qqbot-runtime/src/main/kotlin/com/mieai/qqbot/runtime/security/KeyMaterial.kt
package com.mieai.qqbot.runtime.security

import java.security.spec.AlgorithmParameterSpec
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

/** Identified AES-256 master key supplied by a [KeyProvider]. */
class KeyMaterial(
    val keyId: String,
    key: SecretKey,
) {
    private val key: SecretKey

    init {
        requireKeyId(keyId)
        require(key.algorithm.equals("AES", ignoreCase = true)) { "key algorithm must be AES" }
        val encoded = key.encoded ?: throw IllegalArgumentException("key must expose RAW key material")
        try {
            require(encoded.size == AES_256_BYTES) { "key must contain exactly 256 bits" }
            this.key = SecretKeySpec(encoded, "AES")
        } finally {
            encoded.fill(0)
        }
    }

    internal fun initializeCipher(
        cipher: Cipher,
        operationMode: Int,
        parameters: AlgorithmParameterSpec,
    ) {
        cipher.init(operationMode, key, parameters)
    }

    override fun toString(): String = "KeyMaterial[keyId=$keyId, key=<redacted>]"

    private companion object {
        const val AES_256_BYTES = 32

        fun requireKeyId(value: String): String {
            require(value.isNotBlank()) { "keyId must not be blank" }
            require(value == value.trim()) { "keyId must not have surrounding whitespace" }
            require(value.codePoints().noneMatch(Character::isWhitespace)) {
                "keyId must not contain whitespace"
            }
            require(value.codePoints().noneMatch(Character::isISOControl)) {
                "keyId must not contain control characters"
            }
            return value
        }
    }
}
